README.md
# tiny-led-matrix

A Rust library for direct control of a small monochrome LED display.

This is designed to support the [micro:bit](https://microbit.org/)'s 5×5
display, but nothing in this crate is specific to the micro:bit.


## Documentation

https://docs.rs/tiny-led-matrix


## Changelog

See `CHANGELOG.md`.


## Licence

Copyright 2019 Matthew Woodcraft.

Licensed under either of

 * Apache Licence, Version 2.0
   ([LICENSE-APACHE](LICENSE-APACHE) or http://www.apache.org/licenses/LICENSE-2.0)
 * MIT licence
   ([LICENSE-MIT](LICENSE-MIT) or http://opensource.org/licenses/MIT)

at your option.


## Contribution

Unless you explicitly state otherwise, any contribution intentionally submitted
for inclusion in the work by you, as defined in the Apache-2.0 licence, shall be
dual licensed as above, without any additional terms or conditions.



CHANGELOG.md
## tiny-led-matrix 1.0.2 (2020-01-07)

* Fix repository URL in Cargo.toml


## tiny-led-matrix 1.0.1 (2019-11-16)

* Update examples in docs to use cortex-m-rtfm v0.5


## tiny-led-matrix 1.0.0 (2019-04-27)

* Make `Display::handle_event()`` return a value indicating whether a new
  primary cycle has begun.


## tiny-led-matrix 0.1.0 (2019-03-10)

* Initial release



src/lib.rs
#![no_std]

//! A library for direct control of a small monochrome LED display.
//!
//! This is designed to support the [micro:bit](https://microbit.org/)'s 5×5
//! display, but nothing in this crate is specific to the micro:bit.
//!
//! The library assumes the display is internally organised as a two-dimensional
//! matrix of rows and columns, with the individual LEDs addressed directly
//! (typically with one GPIO pin for each row and one for each column).
//!
//! It isn't designed for the kind of display where you “clock in” the data
//! for an LED row using a smaller number of output pins.
//!
//!
//! # Display model
//!
//! The LED display must be organised as a two-dimensional array of *matrix
//! rows* and *matrix columns*. These describe how the LEDs are wired up, and
//! need not match their visible arrangement.
//!
//! At any time, LEDs from at most one matrix row of the display are lit; the
//! display driver repeatedly lights LEDs from each row in turn to create the
//! illusion of a stable image.
//!
//! At present this crate supports at most 16 matrix columns. There's no
//! strict limit on the number of rows, but in practice it must be small so
//! that the LEDs are lit for a reasonable proportion of the time.
//!
//!
//! ## Greyscale model
//!
//! LED brightness levels are described using a scale from 0 (off) to 9
//! (brightest) inclusive.
//!
//! These are converted to time slices using the same relative durations as
//! the [micro:bit MicroPython port][micropython] uses.
//!
//! The time slice for each level above 1 is approximately 1.9× the slice for
//! the previous level.
//!
//! If there are three matrix rows in the display, an LED with brightness 9 is
//! lit for one third of the time.
//!
//!
//! # Configuring the library for a device.
//!
//! To use this library, you will have to supply implementations of a number
//! of traits, describing your device and its display.
//!
//!
//! ## Matrix
//!
//! You must supply an implementation of the [`Matrix`] trait to describe the
//! matrix dimensions, and how the matrix corresponds to the visible arrangement
//! of LEDs.
//!
//!
//! ## Images and Render
//!
//! The [`Render`] trait defines the interface that an image-like type needs to
//! provide in order to be displayed.
//!
//! The image reports the brightness to use for a given LED, given coordinates
//! according to the visible arrangement.
//!
//! This crate doesn't supply any implementations of `Render`; you should
//! define at least one image type of a suitable size and implement `Render`
//! for it.
//!
//!
//! ## Frames
//!
//! Types implementing [`Render`] are used to update a [`Frame`] (which is in
//! turn passed to a [`Display`]).
//!
//! A `Frame` instance is a 'compiled' representation of a greyscale image of
//! the size required by the display, in a form that's more directly usable by
//! the display code.
//!
//! This is exposed in the public API so that you can construct the `Frame`
//! representation in code running at a low priority. Then only
//! [`Display::set_frame()`] has to be called in code that can't be interrupted
//! by the display timer.
//!
//! You must supply an implementation of the [`Frame`] trait.
//!
//!
//! ## Timer control
//!
//! The `Display` expects to control a single timer which will generate
//! interrupts at appropriate times. [`Display::handle_event()`] is intended to
//! be called from these interrupts.
//!
//! You must supply an implementation of the [`DisplayTimer`] trait providing
//! the interface that the `Display` needs to control the timer.
//!
//! The [`DisplayTimer`] implementation determines the refresh rate for the
//! display.
//!
//! The `Display` requests an interrupt for the point in time when the next
//! row is due to be lit. When rendering greyscale images, it requests
//! additional interrupts within each row's time period. It only requests
//! interrupts for the greyscale levels which are required for what's
//! currently being displayed.
//!
//!
//! ## LED control
//!
//! The `Display` expects to be able to light an arbitrary subset of the LEDs in
//! a given matrix row.
//!
//! You must supply an implementation of the [`DisplayControl`] trait to provide
//! the interface that it needs.
//!
//!
//! # Using the library
//!
//! ## Display
//!
//! A [`Display`] instance controls the LEDs and programs a timer. There will
//! normally be a single `Display` instance in a program using this library.
//!
//! `Display` is generic over a type implementing [`Frame`], which in turn
//! determines the [`Matrix`] in use.
//!
//! ## Putting it together
//!
//! Once you have provided implementations of all the necessary traits, you
//! can use this library as follows:
//!
//! When your program starts, call [`initialise_control()`] (passing it the
//! device implementing `DisplayControl`) and [`initialise_timer()`] (passing
//! it the device implementing `DisplayTimer`), and create a [`Display`] using
//! your [`Frame`] type.
//!
//! In an interrupt handler for the timer you used for `initialise_timer()`,
//! call [`Display::handle_event()`], passing it the same two devices.
//!
//! To display an image: create a [`Frame`] instance, use [`Frame::set()`] to
//! put the image in it, then call [`Display::set_frame()`].
//!
//! You can call `set_frame()` at any time, so long as you're not
//! interrupting, or interruptable by, `handle_event()`.
//!
//! Once you've called `set_frame()`, you are free to reuse the `Frame`
//! instance.
//!
//!
//! [micropython]: https://microbit-micropython.readthedocs.io/


mod control;
mod display;
mod timer;
mod render;

pub use control::DisplayControl;
pub use display::{RowPlan, Matrix, Frame, Display, Event,
                  initialise_timer, initialise_control,
};
pub use timer::DisplayTimer;
pub use render::{BRIGHTNESSES, MAX_BRIGHTNESS, Render};


src/timer.rs
//! The interface that [`Display`] needs to work with a timer.
//!
//! [`Display`]: crate::display::Display


/// The interface that [`Display`] needs to work with a timer.
///
/// The timer should count _ticks_ of a fixed size.
///
/// It should reset itself after the number of ticks passed to
/// `initialise_cycle()` have elapsed (the _primary cycle_), and signal an
/// interrupt.
///
/// It should also provide a _secondary alarm_ which can be programmed to
/// signal an interrupt at a specified point during the primary cycle.
///
/// If you provide a 'no-op' implementation of the secondary-alarm features,
/// the effect will be a display which treats brightness level 9 as on and all
/// other levels as off.
///
/// # Choosing the tick length
///
/// The cycle length is in fact always 375 ticks, so the display will be fully
/// refreshed after `MATRIX_ROWS`×375 ticks. Aiming for around 60 refreshes
/// per second is common.
///
/// For example, if there are three matrix rows you might use a tick period of
/// 16µs. Then the display will be refreshed every 18ms, or about 55 times a
/// second (this is the refresh rate used for the [micro:bit runtime][dal] and
/// the [micro:bit MicroPython port][micropython]).
///
///
/// [dal]: https://lancaster-university.github.io/microbit-docs/
/// [micropython]: https://microbit-micropython.readthedocs.io/
/// [`Display`]: crate::display::Display

pub trait DisplayTimer {

    /// Initialises the timer.
    ///
    /// This is intended to be called once, before using the display.
    ///
    /// The `ticks` parameter is the number of ticks in the primary cycle.
    ///
    /// Leaves the secondary alarm disabled.
    fn initialise_cycle(&mut self, ticks: u16);

    /// Enables the secondary alarm.
    ///
    /// After this is called, an interrupt should be generated each time the
    /// tick last passed to [`program_secondary()`]is reached.
    ///
    /// [`program_secondary()`]: DisplayTimer::program_secondary
    fn enable_secondary(&mut self);

    /// Disables the secondary alarm.
    ///
    /// After this is called, no more interrupts should be generated from the
    /// secondary alarm until it is enabled again.
    fn disable_secondary(&mut self);

    /// Specifies the tick to use for the secondary alarm.
    ///
    /// Note that `ticks` represents the time after the start of the primary
    /// cycle (not the time after the previous secondary signal).
    fn program_secondary(&mut self, ticks: u16);

    /// Checks whether a new primary cycle has begun since the last call to
    /// this method.
    fn check_primary(&mut self) -> bool;

    /// Checks whether the secondary alarm has signalled an interrupt since
    /// the last call to this method.
    fn check_secondary(&mut self) -> bool;

}



src/render.rs
//! The interface between images and the display.

/// The number of brightness levels for greyscale images (ie, 10).
pub const BRIGHTNESSES : usize = 10;

/// The maximum brightness level for greyscale images (ie, 9; the minimum is 0).
pub const MAX_BRIGHTNESS : u8 = (BRIGHTNESSES as u8)-1;


/// A trait providing the information that [`Display`] needs to render an image.
///
/// [`Display`]: crate::display::Display
///
/// # Example implementation
///
/// Here's an example of a 5×5 image type, using one byte for each 'pixel':
///
/// ```
/// # use tiny_led_matrix::Render;
/// struct GreyscaleImage (
///     [[u8; 5]; 5]
/// );
///
/// impl GreyscaleImage {
///     const fn new(data: &[[u8; 5]; 5]) -> GreyscaleImage {
///         GreyscaleImage(*data)
///     }
/// }
///
/// impl Render for GreyscaleImage {
///     fn brightness_at(&self, x: usize, y: usize) -> u8 {
///         self.0[y][x]
///     }
/// }
/// const GREY_HEART: GreyscaleImage = GreyscaleImage::new(&[
///     [0, 9, 0, 9, 0],
///     [9, 5, 9, 5, 9],
///     [9, 5, 5, 5, 9],
///     [0, 9, 5, 9, 0],
///     [0, 0, 9, 0, 0],
/// ]);
/// ```


pub trait Render {

    /// Returns the brightness value for a single LED.
    ///
    /// The ranges for the x and y coordinates are 0..IMAGE_COLS and
    /// 0..IMAGE_ROWS, as defined by the [`Matrix`] for the [`Display`]'s
    /// [`Frame`].
    ///
    /// (0, 0) is the top left.
    ///
    /// The result must be in the range 0..=[`MAX_BRIGHTNESS`]
    ///
    /// # Panics
    ///
    /// If the provided coordinates are out of range, may panic or return an
    /// arbitrary in-range result.
    ///
    /// [`Display`]: crate::display::Display
    /// [`Matrix`]: crate::display::Matrix
    /// [`Frame`]: crate::display::Frame
    fn brightness_at(&self, x: usize, y: usize) -> u8;

}



src/control.rs
//! The interface that [`Display`] needs to control an LED matrix.
//!
//! [`Display`]: crate::display::Display


/// The interface that [`Display`] needs to work with an LED matrix.
///
/// Assumes the matrix is organised by rows and columns, in such a way that
/// LEDs from at most one row are lit at any time.
///
/// [`Display`]: crate::display::Display

pub trait DisplayControl {

    /// Performs any required hardware initialisation.
    ///
    /// This is intended to be called once, before using a display with this
    /// DisplayControl.
    fn initialise_for_display(&mut self);

    /// Lights LEDs in a single matrix row.
    ///
    /// In the specified row, lights exactly the LEDs listed in `cols`.
    /// Turns off all LEDs in the other matrix rows.
    ///
    /// In `cols`, the least-significant bit represents column 0, and so on.
    fn display_row_leds(&mut self, row: usize, cols: u32);

    /// Lights additional LEDs in the current matrix row.
    ///
    /// Affects the row most recently passed to `display_row_leds()`.
    /// Lights the LEDs listed in `cols`, in addition to any already lit.
    ///
    /// In `cols`, the least-significant bit represents column 0, and so on.
    fn light_current_row_leds(&mut self, cols: u32);

}



src/display.rs
//! The display driver and Frame trait.

use crate::control::DisplayControl;
use crate::timer::DisplayTimer;
use crate::render::{Render, BRIGHTNESSES, MAX_BRIGHTNESS};


/// A set of matrix column indices.
///
/// Supports maximum index 15.
#[derive(Copy, Clone, Debug)]
struct ColumnSet (u16);

impl ColumnSet {

    /// Returns a new empty set.
    const fn empty() -> ColumnSet {
        ColumnSet(0)
    }

    /// Adds column index 'col' to the set.
    fn set(&mut self, col: usize) {
        self.0 |= 1<<col;
    }

    /// Returns the set as a bitmap in a u32 (LSB is index 0).
    fn as_u32(&self) -> u32 {
        self.0 as u32
    }

    /// Says whether the set is empty.
    fn is_empty(&self) -> bool {
        self.0 == 0
    }
}


/// A 'compiled' representation of the part of an image displayed on a single
/// matrix row.
///
/// RowPlans are created and contained by [`Frame`]s.
// This is effectively a map brightness -> set of columns
#[derive(Copy, Clone, Debug)]
pub struct RowPlan (
    [ColumnSet; BRIGHTNESSES],
);

impl RowPlan {

    /// Returns a new RowPlan with all LEDs brightness 0.
    pub const fn default() -> RowPlan {
        RowPlan([ColumnSet::empty(); BRIGHTNESSES])
    }

    /// Resets all LEDs to brightness 0.
    fn clear(&mut self) {
        self.0 = RowPlan::default().0;
    }

    /// Says which LEDs have the specified brightness.
    fn lit_cols(&self, brightness: u8) -> ColumnSet {
        self.0[brightness as usize]
    }

    /// Sets a single LED to the specified brightness.
    fn light_col(&mut self, brightness: u8, col: usize) {
        self.0[brightness as usize].set(col);
    }

}


/// Description of a device's LED layout.
///
/// This describes the correspondence between the visible layout of LEDs and
/// the pins controlling them.
///
/// # Example implementation
///
/// ```
/// # use tiny_led_matrix::Matrix;
/// struct SimpleMatrix ();
///
/// impl Matrix for SimpleMatrix {
///     const MATRIX_COLS: usize = 2;
///     const MATRIX_ROWS: usize = 3;
///     const IMAGE_COLS: usize = 3;
///     const IMAGE_ROWS: usize = 2;
///     fn image_coordinates(col: usize, row: usize) -> Option<(usize, usize)> {
///         Some((row, col))
///     }
/// }
/// ```
pub trait Matrix {
    /// The number of pins connected to LED columns.
    ///
    /// At present this can be at most 16.
    const MATRIX_COLS: usize;

    /// The number of pins connected to LED rows.
    ///
    /// This should normally be a small number (eg 3).
    const MATRIX_ROWS: usize;

    // Note that nothing uses IMAGE_COLS and IMAGE_ROWS directly; having these
    // constants allows us to document them.

    /// The number of visible LED columns.
    const IMAGE_COLS: usize;

    /// The number of visible LED rows.
    const IMAGE_ROWS: usize;

    /// Returns the image coordinates (x, y) to use for the LED at (col, row).
    ///
    /// Returns None if (col, row) doesn't control an LED.
    ///
    /// Otherwise the return value is in (0..IMAGE_COLS, 0..IMAGE_ROWS), with
    /// (0, 0) representing the top left.
    ///
    /// # Panics
    ///
    /// Panics if the provided col and row are out of range 0..MATRIX_COLS and
    /// 0..MATRIX_ROWS.

    fn image_coordinates(col: usize, row: usize) -> Option<(usize, usize)>;
}


/// A 'Compiled' representation of an image to be displayed.
///
/// `Frame`s are populated from images implementing [`Render`], then passed on
/// to [`Display::set_frame()`].
///
/// # Implementing `Frame`
///
/// Implementations of `Frame` do two things:
///
/// - specify the [`Matrix`] used to convert between image and matrix
///   coordinates
/// - act like an array of [`RowPlan`]s, one for each matrix row.
///
/// Note that implementations of `Frame` must also implement `Copy` and
/// `Default`.
///
/// # Example implementation
///
/// ```
/// # use tiny_led_matrix::{Matrix,Frame,RowPlan};
/// # struct SimpleMatrix ();
/// # impl Matrix for SimpleMatrix {
/// #     const MATRIX_COLS: usize = 2;
/// #     const MATRIX_ROWS: usize = 3;
/// #     const IMAGE_COLS: usize = 3;
/// #     const IMAGE_ROWS: usize = 2;
/// #     fn image_coordinates(col: usize, row: usize) ->
/// #                         Option<(usize, usize)> {
/// #         Some((row, col))
/// #     }
/// # }
/// #
/// #[derive(Copy, Clone)]
/// struct SimpleFrame (
///     [RowPlan; 3]
/// );
///
/// impl Default for SimpleFrame {
///     fn default() -> SimpleFrame {
///         SimpleFrame([RowPlan::default(); SimpleFrame::ROWS])
///     }
/// }
///
/// impl Frame for SimpleFrame {
///     type Mtx = SimpleMatrix;
///
///     fn row_plan(&self, row: usize) -> &RowPlan {
///         &self.0[row]
///     }
///
///     fn row_plan_mut(&mut self, row: usize) -> &mut RowPlan {
///         &mut self.0[row]
///     }
/// }
/// ```
pub trait Frame: Copy + Default {

    /// The Matrix used to convert between image and matrix coordinates.
    type Mtx: Matrix;

    /// The number of pins connected to LED columns.
    const COLS: usize = Self::Mtx::MATRIX_COLS;

    /// The number of pins connected to LED rows.
    const ROWS: usize = Self::Mtx::MATRIX_ROWS;

    /// Returns a reference to the RowPlan for a row of LEDs.
    ///
    /// # Panics
    ///
    /// Panics if `row` is not in the range 0..ROWS
    fn row_plan(&self, row: usize) -> &RowPlan;

    /// Returns a mutable reference to the RowPlan for a row of LEDs.
    ///
    /// # Panics
    ///
    /// Panics if `row` is not in the range 0..ROWS
    fn row_plan_mut(&mut self, row: usize) -> &mut RowPlan;


    /// Stores a new image into the frame.
    ///
    /// Example:
    ///
    /// ```ignore
    /// frame.set(GreyscaleImage::blank());
    /// ```
    fn set<T>(&mut self, image: &T) where T: Render + ?Sized {
        for row in 0..Self::ROWS {
            let plan = self.row_plan_mut(row);
            plan.clear();
            for col in 0..Self::COLS {
                if let Some((x, y)) = Self::Mtx::image_coordinates(col, row) {
                    let brightness = image.brightness_at(x, y);
                    plan.light_col(brightness, col);
                }
            }
        }
    }
}


// With a 16µs period, 375 ticks is 6ms
const CYCLE_TICKS: u16 = 375;

const GREYSCALE_TIMINGS: [u16; BRIGHTNESSES-2] = [
//   Delay,   Bright, Ticks, Duration, Relative power
//   375,  //   0,      0,      0µs,    ---
     373,  //   1,      2,     32µs,    inf
     371,  //   2,      4,     64µs,   200%
     367,  //   3,      8,    128µs,   200%
     360,  //   4,     15,    240µs,   187%
     347,  //   5,     28,    448µs,   187%
     322,  //   6,     53,    848µs,   189%
     273,  //   7,    102,   1632µs,   192%
     176,  //   8,    199,   3184µs,   195%
//     0,  //   9,    375,   6000µs,   188%
];


/// The reason for a display-timer interrupt.
///
/// This is the return value from [`handle_event()`].
///
/// [`handle_event()`]: Display::handle_event
#[derive(PartialEq, Eq, Debug)]
pub enum Event {
    /// The display has switched to lighting a new row.
    SwitchedRow,
    /// The display has changed the LEDs in the current row.
    UpdatedRow,
    /// Neither a new primary cycle nor a secondary alarm has occurred.
    Unknown,
}

impl Event {
    /// Checks whether this event is `SwitchedRow`.
    ///
    /// This is provided for convenience in the common case where you want to
    /// perform some action based on the display timer's primary cycle.
    pub fn is_new_row(self) -> bool {
        self == Event::SwitchedRow
    }
}


/// Starts the timer you plan to use with a [`Display`].
///
/// Call this once before using a [`Display`].
///
/// This calls the timer's
/// [`initialise_cycle()`][DisplayTimer::initialise_cycle] implementation.
pub fn initialise_timer(timer: &mut impl DisplayTimer) {
    timer.initialise_cycle(CYCLE_TICKS);
}


/// Initialises the display hardware you plan to use with a [`Display`].
///
/// Call this once before using a [`Display`].
///
/// This calls the [`DisplayControl`]'s
/// [`initialise_for_display()`][DisplayControl::initialise_for_display]
/// implementation.
pub fn initialise_control(control: &mut impl DisplayControl) {
    control.initialise_for_display();
}


/// Manages a small LED display.
///
/// There should normally be a single `Display` instance for a single piece of
/// display hardware.
///
/// Display is generic over a [`Frame`] type, which holds image data suitable
/// for display on a particular piece of hardware.
///
/// Call [`initialise_control()`] and [`initialise_timer()`] before using a
/// `Display`.
///
/// # Example
///
/// Using `cortex-m-rtfm` v0.5:
/// ```ignore
/// #[app(device = nrf51, peripherals = true)]
/// const APP: () = {
///
///     struct Resources {
///         gpio: nrf51::GPIO,
///         timer1: nrf51::TIMER1,
///         display: Display<MyFrame>,
///     }
///
///     #[init]
///     fn init(cx: init::Context) -> init::LateResources {
///         let mut p: nrf51::Peripherals = cx.device;
///         display::initialise_control(&mut MyDisplayControl(&mut p.GPIO));
///         display::initialise_timer(&mut MyDisplayTimer(&mut p.TIMER1));
///         init::LateResources {
///             gpio : p.GPIO,
///             timer1 : p.TIMER1,
///             display : Display::new(),
///         }
///     }
/// }
/// ```

pub struct Display<F: Frame> {
    // index (0..F::ROWS) of the row being displayed
    row_strobe      : usize,
    // brightness level (0..=MAX_BRIGHTNESS) to process next
    next_brightness : u8,
    frame           : F,
    current_plan    : RowPlan
}

impl<F: Frame> Display<F> {

    /// Creates a Display instance, initially holding a blank image.
    pub fn new() -> Display<F> {
        Display {
            row_strobe: 0,
            next_brightness: 0,
            frame: F::default(),
            current_plan: RowPlan::default(),
        }
    }

    /// Accepts a new image to be displayed.
    ///
    /// The code that calls this method must not be interrupting, or
    /// interruptable by, [`handle_event()`][Display::handle_event].
    ///
    /// After calling this, it's safe to modify the frame again (its data is
    /// copied into the `Display`).
    ///
    /// # Example
    ///
    /// In the style of `cortex-m-rtfm` v0.5:
    ///
    /// ```ignore
    /// #[task(binds = RTC0, priority = 1, resources = [rtc0, display])]
    /// fn rtc0(mut cx: rtc0::Context) {
    ///     static mut FRAME: MyFrame = MyFrame::const_default();
    ///     &cx.resources.rtc0.clear_tick_event();
    ///     FRAME.set(GreyscaleImage::blank());
    ///     cx.resources.display.lock(|display| {
    ///         display.set_frame(FRAME);
    ///     });
    /// }
    /// ```
    pub fn set_frame(&mut self, frame: &F) {
        self.frame = *frame;
    }

    /// Updates the display for the start of a new primary cycle.
    ///
    /// Leaves the timer's secondary alarm enabled iff there are any
    /// intermediate brightnesses in the current image.
    fn render_row(&mut self,
                  control: &mut impl DisplayControl,
                  timer: &mut impl DisplayTimer) {
        assert! (self.row_strobe < F::ROWS);
        self.row_strobe += 1;
        if self.row_strobe == F::ROWS {self.row_strobe = 0};

        let plan = self.frame.row_plan(self.row_strobe);

        let lit_cols = plan.lit_cols(MAX_BRIGHTNESS);
        control.display_row_leds(self.row_strobe, lit_cols.as_u32());

        // We copy this so that we'll continue using it for the rest of this
        // 'tick' even if set_frame() is called part way through
        self.current_plan = *plan;
        self.next_brightness = MAX_BRIGHTNESS;
        self.program_next_brightness(timer);
        if self.next_brightness != 0 {
            timer.enable_secondary();
        }
    }

    /// Updates the display to represent an intermediate brightness.
    ///
    /// This is called after an interrupt from the secondary alarm.
    fn render_subrow(&mut self,
                     control: &mut impl DisplayControl,
                     timer: &mut impl DisplayTimer) {
        // When this method is called, next_brightness is an intermediate
        // brightness in the range 1..8 (the one that it's time to display).

        let additional_cols = self.current_plan.lit_cols(self.next_brightness);
        control.light_current_row_leds(additional_cols.as_u32());

        self.program_next_brightness(timer);
    }

    /// Updates next_brightness to the next (dimmer) brightness that needs
    /// displaying, and program the timer's secondary alarm correspondingly.
    ///
    /// If no further brightness needs displaying for this row, this means
    /// disabling the secondary alarm.
    fn program_next_brightness(&mut self, timer: &mut impl DisplayTimer) {
        loop {
            self.next_brightness -= 1;
            if self.next_brightness == 0 {
                timer.disable_secondary();
                break;
            }
            if !self.current_plan.lit_cols(self.next_brightness).is_empty() {
                timer.program_secondary(
                    GREYSCALE_TIMINGS[(self.next_brightness-1) as usize]
                );
                break;
            }
        }
    }

    /// Updates the LEDs and timer state during a timer interrupt.
    ///
    /// You should call this each time the timer's interrupt is signalled.
    ///
    /// The `timer` parameter must represent the same device each time you
    /// call this method, and the same as originally passed to
    /// [`initialise_timer()`].
    ///
    /// The `control` parameter must represent the same device each time you
    /// call this method, and the same as originally passed to
    /// [`initialise_control()`].
    ///
    /// This method always calls the timer's
    /// [`check_primary()`][DisplayTimer::check_primary] and
    /// [`check_secondary()`][DisplayTimer::check_secondary] methods.
    ///
    /// As well as updating the LED state by calling [`DisplayControl`]
    /// methods, it may update the timer state by calling the timer's
    /// [`program_secondary()`][DisplayTimer::program_secondary],
    /// [`enable_secondary()`][DisplayTimer::enable_secondary], and/or
    /// [`disable_secondary()`][DisplayTimer::disable_secondary] methods.
    ///
    /// Returns a value indicating the reason for the interrupt. You can check
    /// this if you wish to perform some other action once per primary cycle.
    ///
    /// # Example
    ///
    /// In the style of `cortex-m-rtfm` v0.5:
    ///
    /// ```ignore
    /// #[task(binds = TIMER1, priority = 2,
    ///        resources = [timer1, gpio, display])]
    /// fn timer1(cx: timer1::Context) {
    ///     let display_event = cx.resources.display.handle_event(
    ///         &mut MyDisplayControl(&mut cx.resources.timer1),
    ///         &mut MyDisplayControl(&mut cx.resources.gpio),
    ///     );
    ///     if display_event.is_new_row() {
    ///         ...
    ///     }
    /// }
    /// ```
    pub fn handle_event(&mut self,
                        timer: &mut impl DisplayTimer,
                        control: &mut impl DisplayControl) -> Event {
        let row_timer_fired = timer.check_primary();
        let brightness_timer_fired = timer.check_secondary();
        if row_timer_fired {
            self.render_row(control, timer);
            Event::SwitchedRow
        } else if brightness_timer_fired {
            self.render_subrow(control, timer);
            Event::UpdatedRow
        } else {
            Event::Unknown
        }
    }

}

